.circleci/workflows.yml

```diff
@@ -299,6 +303,105 @@ jobs:
+    resource_class: macos.m1.large.gen1 # 8 cores
+    environment:
+      FERROCENE_HOST: aarch64-apple-darwin
+      FERROCENE_TARGETS: aarch64-apple-darwin,x86_64-apple-darwin
+      SCRIPT: ferrocene/ci/scripts/llvm-cache.sh prepare
+      # This tries to strike a balance between the highest concurrency possible
+      # and not running out of memory during a build. If building LLVM fails
+      # due to the OOM killer killing the build, decrease this number.
+      LLVM_BUILD_PARALLELISM: 18
+    steps:
+      - when:
+          condition: << pipeline.parameters.llvm-rebuild--aarch64-apple-darwin >>
+          steps:
+            - ferrocene-checkout
+            - ferrocene-setup-darwin # Darwin does not come with awscli, setup it before aws steps
+            - aws-oidc-auth
+            - ferrocene-ci
+      - run:
+          name: Empty step to make sure the job always has steps
+          command: echo
+
+  aarch64-darwin-build:
+    macos:
+      xcode: "15.3.0"
+    resource_class: macos.m1.large.gen1 # 8 cores
+    environment:
+      FERROCENE_HOST: aarch64-apple-darwin
+      FERROCENE_TARGETS: x86_64-apple-darwin,aarch64-apple-darwin
+      SCRIPT: |
+        ferrocene/ci/scripts/llvm-cache.sh download
+        ./x.py --stage 2 build library src/tools/rustdoc
+        ./x.py --stage 1 build src/tools/rustdoc
+    steps:
+      - ferrocene-job-build:
+          os: darwin
+
+  aarch64-darwin-dist:
+    macos:
+      xcode: "15.3.0"
+    resource_class: macos.m1.large.gen1 # 8 cores
+    environment:
+      FERROCENE_HOST: aarch64-apple-darwin
+      FERROCENE_TARGETS: x86_64-apple-darwin,aarch64-apple-darwin
+      SCRIPT: |
+        # See ferrocene/ci/split-tasks.py for a list of tasks executed by this.
+        ./x.py --stage 2 dist $(ferrocene/ci/split-tasks.py dist)
+    steps:
+      - ferrocene-job-dist:
+          os: darwin
+          restore-from-job: aarch64-darwin-build
+
+  aarch64-darwin-dist-tools:
+    macos:
+      xcode: "15.3.0"
+    resource_class: macos.m1.large.gen1 # 8 cores
+    environment:
+      FERROCENE_HOST: aarch64-apple-darwin
+      FERROCENE_TARGETS: x86_64-apple-darwin,aarch64-apple-darwin
+      SCRIPT: |
+        # See ferrocene/ci/split-tasks.py for a list of tasks executed by this.
+        ./x.py --stage 2 dist $(ferrocene/ci/split-tasks.py dist:tools)
+    steps:
+    - ferrocene-job-dist:
+          os: darwin
+          restore-from-job: aarch64-darwin-build
+
+  aarch64-darwin-dist-targets:
+    macos:
+      xcode: "15.3.0"
+    resource_class: macos.m1.large.gen1 # 8 cores
+    environment:
+      FERROCENE_HOST: aarch64-apple-darwin
+      FERROCENE_TARGETS: aarch64-apple-darwin,x86_64-apple-darwin,aarch64-unknown-none,thumbv7em-none-eabi,thumbv7em-none-eabihf,armv8r-none-eabihf,wasm32-unknown-unknown
+      SCRIPT: |
+        ./x.py --stage 2 dist rust-std
+    steps:
+      - ferrocene-job-dist:
+          os: darwin
+          restore-from-job: aarch64-darwin-build
```

Comment:
We don't need to rebuild targets on the apple builders, they just need to be built once for all host platforms.